### DISCUSSION TITLE
mypy in ci vs pre-commit

### DISCUSSION THREAD

**Discussion by raybellwaves** [pydata/xarray#5307] [General] [answered]
We would like to add typing in xskillscore. @blackary made a PR to add it in using pre-commit (https://github.com/xarray-contrib/xskillscore/pull/320).

For the most part we stay close to xarray for our codebase.

I would like to understand https://github.com/pydata/xarray/pull/5110#issuecomment-819656693 a bit more. i.e. what was the decision to bring it out of pre-commit and in to a separate step in gha https://github.com/pydata/xarray/blob/master/.github/workflows/ci-additional.yaml#L149

**Comment by keewis** [comment, marked as answer]
That's because `pre-commit` will create a new environment (using `venv` / `virtualenv`), which contains only `mypy`. This means that `mypy` can check for internal consistency, but the type hints of dependencies (e.g. `numpy`) won't be used to check compatibility with those (we would need a full environment for that).

`pre-commit` has the `additional_dependencies` setting which will install additional packages into the environment, but that's another list of dependencies which can easily get out-of-sync (and I at least didn't know about it when we added the separate `mypy` CI).
